### PR TITLE
Thin pool creation

### PR DESCRIPTION
**Pull request by sp98**
Create thin pool via VG Manager.

Testing:

CR:
```
apiVersion: lvm.topolvm.io/v1alpha1
kind: LVMCluster
metadata:
  name: lvmcluster-sample
spec:
  storage:
    deviceClasses:
    - name: vg1
      thinPoolConfig:
        name: mytp1
        sizePercent: 50
        overprovisionRatio: 50
```

```
sh-4.4# vgs
  VG  #PV #LV #SN Attr   VSize  VFree
  vg1   3   1   0 wz--n- <4.37t 2.18t
sh-4.4# lvs
  LV    VG  Attr       LSize Pool Origin Data%  Meta%  Move Log Cpy%Sync Convert
  mytp1 vg1 twi-a-tz-- 2.18t             0.00   6.13
sh-4.4#


```

```
sh-4.4# cat lvmd.yaml
device-classes:
- default: true
  lvcreate-options: null
  name: vg2
  spare-gb: null
  stripe: null
  stripe-size: ""
  thin-pool:
    name: mytp2
    overprovision-ratio: 30
  type: thin
  volume-group: vg2
socket-name: /run/lvmd/lvmd.sock
```

Apps using thin volumes:
``` sh-4.4# vgs
  VG  #PV #LV #SN Attr   VSize  VFree
  vg2   3   3   0 wz--n- <4.37t 2.18t
sh-4.4# lvs
  LV                                   VG  Attr       LSize  Pool  Origin Data%  Meta%  Move Log Cpy%Sync Convert
  76e79637-5cc8-4795-8644-4298d8db3d1b vg2 Vwi-aotz--  5.00g mytp2        0.29
  c293d9ab-04dc-4b6d-8d9c-b3bf368a091c vg2 Vwi-aotz-- 40.00g mytp2        0.07
  mytp2                                vg2 twi-aotz--  2.18t              0.01   6.13
sh-4.4#
```

### CODE REVIEW

File: api/v1alpha1/lvmcluster_types.go [hunk 1]
```diff
@@ -36,6 +36,20 @@ type LVMClusterSpec struct {
...
 }
 
+type ThinPoolConfig struct {
+	// Name of the thin pool to be creatd
+	Name string `json:"name,omitempty"`
+
+	// SizePercent represents size percentage of the thin-pool with respect
+	// to the available volume group size. For example: if size is 60 and
+	// available volume group size is 10G, then thin-pool size would be 6G
+	SizePercent int `json:"sizePercent,omitempty"`
```
Comment (nbalacha) [inline api/v1alpha1/lvmcluster_types.go:46]: Please set a default value + a validation check.

File: api/v1alpha1/lvmcluster_types.go [hunk 2]
```diff
@@ -45,6 +59,14 @@ type DeviceClass struct {
 	// +kubebuilder:validation:Pattern="^[a-z0-9]([-a-z0-9]*[a-z0-9])?$"
 	Name string `json:"name,omitempty"`
 
+	// VolumeGroupName is the name of the volume group to be created.
+	// Validations to confirm that this field can be used as metadata.name field in storageclass
+	// ref: https://github.com/kubernetes/apimachinery/blob/de7147/pkg/util/validation/validation.go#L209
+	// +kubebuilder:validation:MaxLength=245
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:Pattern="^[a-z0-9]([-a-z0-9]*[a-z0-9])?$"
+	VolumeGroupName string `json:"volumeGroupName,omitempty"`
```
Comment (nbalacha) [inline api/v1alpha1/lvmcluster_types.go:68]: I'm not convinced that this field is required just yet.

File: api/v1alpha1/lvmvolumegroup_types.go [hunk 3]
```diff
@@ -25,6 +25,9 @@ import (
 
 // LVMVolumeGroupSpec defines the desired state of LVMVolumeGroup
 type LVMVolumeGroupSpec struct {
+	// DeviceClassName is the name of the device class associated with this volume group
+	DeviceClassName string `json:"deviceClassName,omitempty"`
```
Comment (nbalacha) [inline api/v1alpha1/lvmvolumegroup_types.go:29]: I'm not convinced this is required.

File: pkg/vgmanager/vgmanager_controller.go [hunk 4]
```diff
@@ -244,6 +250,29 @@ func (r *VGReconciler) reconcile(ctx context.Context, req ctrl.Request) (ctrl.Re
...
 
+func (r *VGReconciler) addThinPoolToVG(vgName string, config lvmv1alpha1.ThinPoolConfig) error {
+	lvList, err := ListLogicalVolumes(r.executor, vgName)
+	if err != nil {
+		return fmt.Errorf("failed to list logical volumes in the volume group %q. %v", vgName, err)
+	}
+
+	for _, lv := range lvList {
+		if lv == config.Name {
+			r.Log.Info("logical volume is already created", "Volume Group", vgName, "Logical Volume", config.Name)
```
Comment (nbalacha) [inline pkg/vgmanager/vgmanager_controller.go:261]:
```suggestion
			r.Log.Info("lvm thinpool already exists", "vgname", vgName, "thinpool", config.Name)
```

Comment (nbalacha) [reply pkg/vgmanager/vgmanager_controller.go:261]: An additional check to determine whether the existing lv is a thinpool or not would provide more info in the log message.

File: pkg/vgmanager/vgmanager_controller.go [hunk 5]
```diff
@@ -244,6 +250,29 @@ func (r *VGReconciler) reconcile(ctx context.Context, req ctrl.Request) (ctrl.Re
...
+			return nil
+		}
+	}
+
+	args := []string{"-L", "5G", "-c", "512", "-T", fmt.Sprintf("%s/%s", vgName, config.Name)}
```
Comment (nbalacha) [inline pkg/vgmanager/vgmanager_controller.go:266]: Why is the size set to 5G?

File: api/v1alpha1/lvmcluster_types.go (hunk 2 above)
Comment (sp98) [reply api/v1alpha1/lvmcluster_types.go]: removed. Only supporting one pool per device class/ Volume group now.

File: api/v1alpha1/lvmvolumegroup_types.go (hunk 3 above)
Comment (sp98) [reply api/v1alpha1/lvmvolumegroup_types.go]: removed. Only supporting one pool per device class/ Volume group now.

File: pkg/vgmanager/vgmanager_controller.go (hunk 5 above)
Comment (sp98) [reply pkg/vgmanager/vgmanager_controller.go]: using `<x>%FREE` to determine the size, where `x` is `thinPoolConfig.SizePercent`.

File: api/v1alpha1/lvmcluster_types.go (hunk 1 above)
Comment (sp98) [reply api/v1alpha1/lvmcluster_types.go:48]: using below validations:
``` 	// +kubebuilder:validation:default=75
	// +kubebuilder:validation:Minimum=10
	// +kubebuilder:validation:Maximum=90```

File: api/v1alpha1/lvmcluster_types.go [hunk 6]
```diff
@@ -36,6 +36,22 @@ type LVMClusterSpec struct {
...
+	// SizePercent represents percentage of remaining space in the volume group that should be used
+	// for creating the thin pool.
+	// +kubebuilder:validation:default=75
+	// +kubebuilder:validation:Minimum=10
+	// +kubebuilder:validation:Maximum=90
+	SizePercent int `json:"sizePercent,omitempty"`
+
+	// OverProvisionRatio represents the ratio of overprovision that can
+	// be allowed on thin pools
+	OverprovisionRatio int `json:"overprovisionRatio,omitempty"`
```
[the remark this answers is not in the document]
Comment (sp98) [reply api/v1alpha1/lvmcluster_types.go:52]: changed it to `int` after offline discussion.

**Review by leelavg — COMMENTED** [review]
- in general, we don't have any validation of CR yet, better we start adding validation code with this CR or mark mandatory fields as required in CRD via kubebuilder?

**Comment by sp98** [comment]
> * in general, we don't have any validation of CR yet, better we start adding validation code with this CR or mark mandatory fields as required in CRD via kubebuilder?

added CRD validation for the required fields for the new ThinPoolConfig.

**Review by leelavg — APPROVED** [review]
/lgtm

File: go.mod [hunk 7]
```diff
@@ -52,52 +55,51 @@ require (
...
 	k8s.io/component-base v0.23.0 // indirect
-	k8s.io/klog/v2 v2.30.0 // indirect
+	k8s.io/klog/v2 v2.40.1 // indirect
 	k8s.io/kube-openapi v0.0.0-20211115234752-e816edb12b65 // indirect
 	k8s.io/utils v0.0.0-20211116205334-6203023598ed // indirect
 	sigs.k8s.io/json v0.0.0-20211020170558-c049b76a60c6 // indirect
 	sigs.k8s.io/structured-merge-diff/v4 v4.2.1 // indirect
 )
+
+replace github.com/topolvm/topolvm => github.com/topolvm/topolvm v0.10.6-0.20220405084709-a4470881be99
```
Comment (sp98) [inline go.mod:105]: This is debatable!

File: api/v1alpha1/lvmcluster_types.go [hunk 8]
```diff
@@ -36,6 +36,25 @@ type LVMClusterSpec struct {
...
+	Name string `json:"name,omitempty"`
+
+	// SizePercent represents percentage of remaining space in the volume group that should be used
+	// for creating the thin pool.
+	// +kubebuilder:validation:default=75
+	// +kubebuilder:validation:Minimum=10
+	// +kubebuilder:validation:Maximum=90
+	SizePercent int `json:"sizePercent,omitempty"`
+
+	// OverProvisionRatio represents the ratio of overprovision that can
```
[the remark this answers is not in the document]
Comment (sp98) [reply api/v1alpha1/lvmcluster_types.go]: done.

File: api/v1alpha1/lvmcluster_types.go [hunk 9]
```diff
@@ -36,6 +36,25 @@ type LVMClusterSpec struct {
 	Storage Storage `json:"storage,omitempty"`
 }
 
+type ThinPoolConfig struct {
+	// Name of the thin pool to be created
+	// +kubebuilder:validation:Required
+	// +required
+	Name string `json:"name,omitempty"`
```
[the remark this answers is not in the document]
Comment (sp98) [reply api/v1alpha1/lvmcluster_types.go]: No. Removed `omitempty`.

File: go.mod [hunk 10]
```diff
@@ -6,9 +6,12 @@ require (
 	github.com/go-logr/logr v1.2.0
 	github.com/google/go-cmp v0.5.6
 	github.com/onsi/ginkgo v1.16.5
+	github.com/onsi/ginkgo/v2 v2.1.3
```
[the remark this answers is not in the document]
Comment (sp98) [reply go.mod:9]: Its there in the second `require` block in the go.mod in `main` branch.  For me, its getting added to the first `require` block for some reason when I run `go mod tidy` after adding `replace` to use topolvm main branch

``` 
# sapillai @ localhost in ~/go/src/github.com/red-hat-storage/lvm-operator on git:thin-pool x [11:17:41] 

$ go mod tidy    
go: finding module for package github.com/onsi/ginkgo/v2
go: found github.com/onsi/ginkgo/v2 in github.com/onsi/ginkgo/v2 v2.1.3
github.com/red-hat-storage/lvm-operator/pkg/vgmanager imports
	sigs.k8s.io/yaml imports
	gopkg.in/yaml.v2 tested by
	gopkg.in/yaml.v2.test imports
	gopkg.in/check.v1 loaded from gopkg.in/check.v1@v1.0.0-20200227125254-8fa46927fb4f,
	but go 1.16 would select v1.0.0-20201130134442-10cb98267c6c

To upgrade to the versions selected by go 1.16:
	go mod tidy -go=1.16 && go mod tidy -go=1.17
If reproducibility with go 1.16 is not needed:
	go mod tidy -compat=1.17
For other options, see:
	https://golang.org/doc/modules/pruning

# sapillai @ localhost in ~/go/src/github.com/red-hat-storage/lvm-operator on git:thin-pool x [11:17:41] 
$ go mod tidy -compat=1.17
go: finding module for package github.com/onsi/ginkgo/v2
go: found github.com/onsi/ginkgo/v2 in github.com/onsi/ginkgo/v2 v2.1.3



```

File: go.mod (hunk 7 above)
Comment (nbalacha) [reply go.mod:105]: Can this be removed seeing that the thinlv PR is now available in the main topolvm project?

**Review by nbalacha — COMMENTED** [review]
Does anything extra need to be done for the cleanup?

File: go.mod (hunk 7 above)
Comment (sp98) [reply go.mod:105]: This points to the `main` branch only.  I think we need to wait for the next topolvm (with thin volume release) to remove this.

File: api/v1alpha1/lvmcluster_types.go [hunk 11]
```diff
@@ -36,6 +36,25 @@ type LVMClusterSpec struct {
...
+	// for creating the thin pool.
+	// +kubebuilder:validation:default=75
+	// +kubebuilder:validation:Minimum=10
+	// +kubebuilder:validation:Maximum=90
+	SizePercent int `json:"sizePercent,omitempty"`
+
+	// OverProvisionRatio is the factor by which additional storage can be provisioned compared to
+	// the available storage in the thin pool.
+	// +kubebuilder:validation:Minimum=2
+	OverprovisionRatio int `json:"overprovisionRatio,omitempty"`
```
Comment (nbalacha) [inline api/v1alpha1/lvmcluster_types.go:55]: can this be empty?

Comment (sp98) [reply api/v1alpha1/lvmcluster_types.go]: No.  Marked it as required.

**Review by nbalacha — CHANGES_REQUESTED** [review]
The thin pool needs to be deleted before the VG is deleted:
{"level":"error","ts":1650198338.4370134,"logger":"controller.lvmcluster","msg":"Reconciler error","reconciler group":"lvm.topolvm.io","reconciler kind":"LVMCluster","name":"lvmcluster-sample","namespace":"lvm-operator-system","error":"failed cleaning up: lvmvg-manager waiting for all VGs to be deleted","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.1/pkg/internal/controller/controller.go:227"}

File: pkg/vgmanager/vgmanager_controller.go [hunk 12]
```diff
@@ -294,6 +294,15 @@ func (r *VGReconciler) processDelete(ctx context.Context, volumeGroup *lvmv1alph
 		return nil
 	}
 
+	// Delete thin pool
+	if volumeGroup.Spec.ThinPoolConfig != nil {
+		args := []string{"-f", fmt.Sprintf("%s/%s", volumeGroup.Name, volumeGroup.Spec.ThinPoolConfig.Name)}
+		_, err = r.executor.ExecuteCommandWithOutputAsHost(lvRemoveCmd, args...)
+		if err != nil {
+			return fmt.Errorf("failed to delete thin pool %q. %v", volumeGroup.Spec.ThinPoolConfig.Name, err)
```
Comment (sp98) [inline pkg/vgmanager/vgmanager_controller.go:302]: It does not cover the case when the thin pool is already deleted by the user.

Comment (sp98) [reply pkg/vgmanager/vgmanager_controller.go]: updated. Checking if the thin pool exists before deleting it.

File: pkg/vgmanager/vgmanager_controller.go [hunk 13]
```diff
@@ -249,6 +294,25 @@ func (r *VGReconciler) processDelete(ctx context.Context, volumeGroup *lvmv1alph
 		return nil
 	}
 
+	// Delete thin pool
```
Comment (nbalacha) [inline pkg/vgmanager/vgmanager_controller.go:297]: This should be done after checking if the volumegroup exists.

Comment (sp98) [reply pkg/vgmanager/vgmanager_controller.go:306]: That's right. Fixed.

**Review by nbalacha — APPROVED** [review]

**Pull request merged**